tools/tpm2_convert_key.c

```diff
@@ -0,0 +1,414 @@
+    struct key_blob blobs[MAX_KEY_BLOBS];
+};
+
+typedef enum {
+    FMT_PEM,
+    FMT_DER,
+    FMT_INVAL
+} convert_format;
+
+// reads just the algorithm type identifier from the given file and returns it
+// in the correct endianess. error handling included.
+static bool read_alg_type(FILE *f, UINT16 *alg_type) {
+
+    if (fread(alg_type, 2, 1, f) != 1) {
+        LOG_ERR("Could not read algtype from input file: \"%s\".",
+            (feof(f) ? "Premature end of file" : strerror(errno))
+        );
+
+        return false;
+    }
+
+    // convert byte order
+    *alg_type = tpm2_util_ntoh_16(*alg_type);
+
+    return true;
+}
+
+// frees any dynamically allocated data in the given key_data struct
+static void free_keys(struct key_data *data) {
+    size_t i;
+    struct key_blob *blob;
+
+    for (i = 0; i < data->num_blobs; i++) {
+        blob = &(data->blobs[i]);
+        free(blob->data);
+        blob->data = NULL;
+    }
+}
+// parse all found blob parts from the given file and add them to the key_data
+// struct. error handling included.
+// on true return the caller is responsible for freeing data
+static bool read_blobs(FILE *f, struct key_data *data) {
+    const size_t MAX_LENGTH = 8192;
+    struct key_blob blob;
+    bool ret = false;
+
+    // read in a variable number of blobs until EOF or error occurs
+    while (true) {
+
+        if (fread(&blob.length, 2, 1, f) != 1) {
+            if (!feof(f)) {
+                LOG_ERR("Could not read key blob length from input file: \"%s\".",
+                    strerror(errno)
+                );
+
+                goto out;
+            }
+
+            ret = true;
+            break;
+        }
+
+        // correct byte order
+        blob.length = tpm2_util_ntoh_16(blob.length);
+
+        if (blob.length > MAX_LENGTH) {
+            LOG_ERR("Excess length key blob of %d bytes encountered",
+                blob.length
+            );
+
+            goto out;
+        }
+        else if (data->num_blobs == MAX_KEY_BLOBS) {
+            LOG_ERR("Maximum number of key blobs %d exceeded.", MAX_KEY_BLOBS);
+            goto out;
+        }
+
+        blob.data = malloc(blob.length);
+
+        if (fread(blob.data, blob.length, 1, f) != 1) {
+            LOG_ERR("Failed to read key blob nr. %zd: \"%s\".",
+                data->num_blobs + 1,
+                (feof(f) ? "Premature end of file" : strerror(errno))
+            );
+
+            free(blob.data);
+
+            goto out;
+        }
+
+        data->blobs[data->num_blobs] = blob;
+        data->num_blobs ++;
+    }
+
+out:
+
+    if (ret == false) {
+        free_keys(data);
+    }
+
+    return ret;
+}
+
+/*
+ * Read the binary structure from infile and put it into the given blob output
+ * parameter. This expects the structure as output by tpm2_akparse.
+ */
+static bool read_key(const char *infile, struct key_data *data) {
+    FILE *f;
+    bool ret = false;
+
+    memset(data, 0, sizeof(struct key_data));
```

Comment:
Any reason you just don't initialize this to 0 when it's declared on the stack?

Reply:
Because then the function would rely on how it's called instead of having defined behaviour by design. This way even if a caller fails to initialize the buffer correctly and to interpret the function success indication we'll have a defined execution flow.